File: game.of.life/game.of.life.domain/Cell.cs
namespace game.of.life.domain
{
    using System.Linq;
    using Newtonsoft.Json;

    public class Cell
    {
        private readonly IComputeNextMutation[] _computeNextMutations =
        {
            new ShouldResurectNextMutation(),
            new ShouldStayAliveNextMutation(),
            new ShouldDieNextMutation()
        };

        public Cell(int x, int y, CellState state = CellState.Dead)
            : this(x, y, state, CellState.Unknown)
        {
        }

        [JsonConstructor]
        public Cell(int x, int y, CellState state, CellState nextState)
        {
            X = x;
            Y = y;
            State = state;
            NextState = nextState;
        }

        public int X { get; }

        public int Y { get; }

        public CellState State { get; private set; }

        public CellState NextState { get; private set; }

        public bool IsAlive()
        {
            return State == CellState.Alive;
        }
        
        /*
        public void ComputeNextMutation(int aliveNeighboursCount)
        {
            if (ShouldResurect(aliveNeighboursCount))
            {
                NextState = CellState.Alive;
            }

            if (ShouldStayAlive(aliveNeighboursCount))
            {
                NextState = CellState.Alive;
            }

            if (ShouldDie(aliveNeighboursCount))
            {
                NextState = CellState.Dead;
            }
        }
        */

        public void ComputeNextMutation(int aliveNeighboursCount)
        {
            _computeNextMutations
                .FirstOrDefault(m => m.CanComputeNextMutation(this, aliveNeighboursCount))
                ?.ComputeNextMutation(this);
        }

        private interface IComputeNextMutation
        {
            bool CanComputeNextMutation(Cell cell, int aliveNeighboursCount);

            void ComputeNextMutation(Cell cell);
        }

        private class ShouldResurectNextMutation : IComputeNextMutation
        {
            public bool CanComputeNextMutation(Cell cell, int aliveNeighboursCount)
            {
                return cell.State == CellState.Dead && aliveNeighboursCount == 3;
            }

            public void ComputeNextMutation(Cell cell)
            {
                cell.NextState = CellState.Alive;
            }
        }

        private class ShouldStayAliveNextMutation : IComputeNextMutation
        {
            public bool CanComputeNextMutation(Cell cell, int aliveNeighboursCount)
            {
                return cell.State == CellState.Alive && (aliveNeighboursCount == 2 || aliveNeighboursCount == 3);
            }

            public void ComputeNextMutation(Cell cell)
            {
                cell.NextState = CellState.Alive;
            }
        }

        private class ShouldDieNextMutation : IComputeNextMutation
        {
            public bool CanComputeNextMutation(Cell cell, int aliveNeighboursCount)
            {
                return cell.State == CellState.Alive && (aliveNeighboursCount < 2 || aliveNeighboursCount >= 4);
            }

            public void ComputeNextMutation(Cell cell)
            {
                cell.NextState = CellState.Dead;
            }
        }

        public void CompleteMutation()
        {
            if (NextState == CellState.Unknown)
            {
                return;
            }

            State = NextState;
            NextState = CellState.Unknown;
        }

        public override bool Equals(object obj)
        {
            if (ReferenceEquals(null, obj))
            {
                return false;
            }

            if (ReferenceEquals(this, obj))
            {
                return true;
            }

            return obj.GetType() == GetType() && Equals((Cell)obj);
        }

        private bool Equals(Cell other)
        {
            return X == other.X && Y == other.Y;
        }

        public override int GetHashCode()
        {
            unchecked
            {
                return (X * 397) ^ Y;
            }
        }

        public override string ToString()
        {
            return $"({X},{Y})";
        }
    }
}

File: game.of.life/game.of.life.ts.web/scripts/GoL/Cell.ts
/// <reference path="../../typings/linq.d.ts"/>

namespace GoL {
	export class Cell {
		private _computeNextMutations: IComputeNextMutation[] =
			[
				new ShouldResurectNextMutation(),
				new ShouldStayAliveNextMutation(),
				new ShouldDieNextMutation()
			];

		public X: number;
		public Y: number;
		public State: CellState;
		public NextState: CellState;

		constructor(x: number, y: number, state: CellState = CellState.Dead, nextState: CellState = CellState.Unknown) {
			this.X = x;
			this.Y = y;
			this.State = state;
			this.NextState = nextState;
		}

		public IsAlive(): boolean {
			return this.State === CellState.Alive;
		}

		public ComputeNextMutation(aliveNeighboursCount: number): void {
			const nextComputation = Enumerable.from(this._computeNextMutations)
				.firstOrDefault((m: IComputeNextMutation) => m.CanComputeNextMutation(this, aliveNeighboursCount));
			
			if (nextComputation) {
				nextComputation.ComputeNextMutation(this);
			}
		}

		public CompleteMutation(): void {
			if (this.NextState === CellState.Unknown) {
				return;
			}

			this.State = this.NextState;
			this.NextState = CellState.Unknown;
		}

		public static equals(a: Cell, b: Cell): boolean {
			if (!a || !b) {
				return false;
			}

			return a.X === b.X && a.Y === b.Y;
		}
		
		public toString(): string
		{
			return `(${this.X},${this.Y})`;
		}
	}

	interface IComputeNextMutation {
		CanComputeNextMutation(cell: GoL.Cell, aliveNeighboursCount: number): boolean;

		ComputeNextMutation(cell: GoL.Cell): void;
	}

	class ShouldResurectNextMutation implements IComputeNextMutation {
		public CanComputeNextMutation(cell: GoL.Cell, aliveNeighboursCount: number): boolean {
			return cell.State === GoL.CellState.Dead && aliveNeighboursCount === 3;
		}

		public ComputeNextMutation(cell: GoL.Cell): void {
			cell.NextState = GoL.CellState.Alive;
		}
	}

	class ShouldStayAliveNextMutation implements IComputeNextMutation {
		public CanComputeNextMutation(cell: GoL.Cell, aliveNeighboursCount: number): boolean {
			return cell.State === GoL.CellState.Alive && (aliveNeighboursCount === 2 || aliveNeighboursCount === 3);
		}

		public ComputeNextMutation(cell: GoL.Cell): void {
			cell.NextState = GoL.CellState.Alive;
		}
	}

	class ShouldDieNextMutation implements IComputeNextMutation {
		public CanComputeNextMutation(cell: GoL.Cell, aliveNeighboursCount: number): boolean {
			return cell.State === GoL.CellState.Alive && (aliveNeighboursCount < 2 || aliveNeighboursCount >= 4);
		}

		public ComputeNextMutation(cell: GoL.Cell): void {
			cell.NextState = GoL.CellState.Dead;
		}
	}
}
